pageserver/src/disk_usage_eviction_task.rs

```diff
@@ -608,6 +640,60 @@ async fn collect_eviction_candidates(
         }
     }
 
+    // FIXME: this is a long loop over all secondary locations.  At the least, respect
+    // cancellation here, but really we need to break up the loop.  We could extract the
+    // Arc<SecondaryTenant>s and iterate over them with some tokio yields in there.  Ideally
+    // though we should just reduce the total amount of work: our eviction goals do not require
+    // listing absolutely every layer in every tenant: we could sample this.
+    let mut secondary_tenants = Vec::new();
+    tenant_manager.foreach_secondary_tenants(
+        |tenant_shard_id: &TenantShardId, state: &Arc<SecondaryTenant>| {
+            secondary_tenants.push((*tenant_shard_id, state.clone()));
+        },
+    );
```

Comment:
Did you consider that a tenant that we already visited in above loop might have transitioned into secondary state by the time we reach here => its layers could be included twice in `candidates`.

Yet another argument to better just carry the `struct Layer` forward between primary and secondary mode.

Reply:
Punting the eviction to a future PR